proposals/2024/10-01_projektor_integration.md

```diff
@@ -0,0 +1,232 @@
+
+4. Cancelled: While or before implementation the proposal was cancelled.
+
+NOTE: This can happen for a multitude of reasons.
+
+5. Complete: This feature/change is implemented.
+-->
+
+## Background
+
+<!--
+This section is intended to describe the new feature, redesign or refactor.
+-->
+
+**Please provide a summary of the new feature, redesign or refactor:**
+
+<!--
+Provide your description here.
+-->
+
+* Currently, there is a lack of native support for handling, visualizing, and tracking test results across builds. Inspired by the feature set in Projektor.dev, this proposal aims to add a dedicated, native support `test-report` feature to Vela.This feature will allow users to parse, store, and visualize test results in a more user-friendly manner.
+
+**Please briefly answer the following questions:**
+
+1. Why is this required?
+
+* Test reporting is a widely used feature in CI/CD pipelines, providing insights into test results, code quality, and performance metrics. By integrating Projektor's test reporting features natively into Vela, users can easily track and analyze test data across builds.
+
+2. If this is a redesign or refactor, what issues exist in the current implementation?
+
+* Currently, there is no native support for handling, visualizing, and tracking test results across builds.
+
+3. Are there any other workarounds, and if so, what are the drawbacks?
+
+* Yes. Users can set up separate infra and interact with it using a plugin to upload and access test reports. However, this is not a native solution and requires additional setup.
+
+4. Are there any related issues? Please provide them below if any exist.
+
+* https://github.com/go-vela/community/issues/528
+
+## Design
+
+<!--
+This section is intended to explain the solution design for the proposal.
+
+NOTE: If there are no current plans for a solution, please leave this section blank.
+-->
+
+**Please describe your solution to the proposal. This includes, but is not limited to:**
+
+* Dedicated `test-report` step.
+* Backend and UI enhancements to visualize test results.
+* Object Storage integration for storing test results.
+* Slack integration for real-time notifications of test results.
+
+
+### 1. Test Report Step Configuration
+A dedicated `test-report` step will be added at the end of the Vela pipeline. This step uses a specialized Docker image (`vela/test-report-agent:latest`) to handle parsing and reporting tasks. Users can define the format, file path, and retention period for test data within this step, ensuring flexibility for different testing frameworks and workflows.
+#### Example Configuration in `.vela.yml`
+```yaml
+steps:
+  - name: test
+    image: golang:latest
+    commands:
+      - go test ./... -json > test-results.json
+  - name: test-report
+    image: vela/test-report-agent:latest
+    parameters:
+      report_format: json
+      report_path: "./test-results.json"
+      retention_days: 30
+      notify_slack: true
+```
```

Comment:
this could be one solution. are there other solutions to consider? can we make it easier for users?

Reply:
We could potentially leverage container outputs enhance to do something like below
```steps:
  - name: test
    image: golang:latest
    commands:
      - go test ./... -json > vela/outputs/test-results.json
  - name: read test report from outputs
    image: ${IMAGE}       # resolves to `ubuntu`
    pull: on_start
    parameters:
      report_format: json
      report_path: "./test-results.json"